File: notale-v2/experiments/workflow-skills-next/sample-workbench/mini/browser-check.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import { chromium } from '/tmp/notale-playwright/node_modules/playwright/index.mjs';

const here=path.dirname(fileURLToPath(import.meta.url));
const manifest=JSON.parse(fs.readFileSync(path.join(here,'manifest.json'),'utf8'));
const baseArgument=process.argv.find(argument=>argument.startsWith('--base-url='));
const sampleArgument=process.argv.find(argument=>argument.startsWith('--sample='));
const baseUrl=(baseArgument?.split('=')[1] || 'http://127.0.0.1:41031').replace(/\/$/,'');
const requested=sampleArgument?.slice('--sample='.length);
const required=manifest.samples.filter(sample=>
  sample.decision==='required' &&
  sample.status==='promoted' &&
  (!requested || sample.key===requested) &&
  fs.existsSync(path.resolve(here,sample.mini,'index.html')) &&
  fs.existsSync(path.join(here,sample.key,'report.json'))
);

if(requested && !required.length){
  console.error(`No promoted mini for ${requested}`);
  process.exit(2);
}

const browser=await chromium.launch();
const results=[];

for(const sample of required){
  for(const profile of [
    {name:'1600-normal',viewport:{width:1600,height:900},reducedMotion:'no-preference'},
    {name:'1280-reduced',viewport:{width:1280,height:720},reducedMotion:'reduce'},
    {
      name:'1600-to-1280-resize',
      viewport:{width:1600,height:900},
      resizeTo:{width:1280,height:720},
      reducedMotion:'no-preference'
    }
  ]){
    const context=await browser.newContext({viewport:profile.viewport,reducedMotion:profile.reducedMotion});
    const page=await context.newPage();
    const errors=[];
    const external=[];
    page.on('pageerror',error=>errors.push(`pageerror: ${error.message}`));
    page.on('console',message=>{
      if(message.type()==='error') errors.push(`console: ${message.text()}`);
    });
    page.on('request',request=>{
      const url=request.url();
      if(/^https?:/.test(url) && !url.startsWith(baseUrl)) external.push(url);
    });
    page.on('requestfailed',request=>errors.push(`requestfailed: ${request.url()}`));
    page.on('response',response=>{
      if(response.status()>=400) errors.push(`HTTP ${response.status()}: ${response.url()}`);
    });

    const url=new URL(sample.mini,`${baseUrl}/sample-workbench/mini/`).href;
    let response;
    try{
      response=await page.goto(url,{waitUntil:'networkidle',timeout:15000});
      await page.waitForTimeout(profile.reducedMotion==='reduce'?150:1400);
      if(profile.resizeTo){
        await page.setViewportSize(profile.resizeTo);
        await page.waitForTimeout(180);
      }
    }catch(error){
      errors.push(`navigation: ${error.message}`);
    }
    const geometry=await page.evaluate(()=>{
      const stage=document.querySelector('#stage');
      const rect=stage?.getBoundingClientRect();
      return {
        viewport:[innerWidth,innerHeight],
        document:[document.documentElement.scrollWidth,document.documentElement.scrollHeight],
        stage:rect?[Math.round(rect.x),Math.round(rect.y),Math.round(rect.width),Math.round(rect.height)]:null,
        canvases:document.querySelectorAll('canvas').length,
        svgs:document.querySelectorAll('svg').length,
        title:document.title
      };
    }).catch(()=>null);
    if(!response || response.status()!==200) errors.push(`HTTP ${response?.status() ?? 'none'}`);
    const expected=profile.resizeTo || profile.viewport;
    if(geometry?.stage && (
      Math.abs(geometry.stage[0])>1 || Math.abs(geometry.stage[1])>1 ||
      Math.abs(geometry.stage[2]-expected.width)>1 ||
      Math.abs(geometry.stage[3]-expected.height)>1
    )) errors.push(`stage geometry ${geometry.stage.join(',')}`);
    if(geometry && (geometry.document[0]>expected.width+1 || geometry.document[1]>expected.height+1)){
      errors.push(`overflow ${geometry.document.join('x')}`);
    }
    if(external.length) errors.push(`external requests: ${[...new Set(external)].join(', ')}`);

    const shots=path.join(here,sample.key,'shots','smoke');
    fs.mkdirSync(shots,{recursive:true});
    const screenshot=path.join(shots,`${profile.name}.png`);
    if(!errors.some(error=>error.startsWith('navigation'))) await page.screenshot({path:screenshot});
    results.push({sample:sample.key,profile:profile.name,status:errors.length?'fail':'pass',errors,geometry,screenshot});
    await context.close();
  }
}

await browser.close();
const failures=results.filter(result=>result.status==='fail');
for(const result of results){
  console.log(`${result.status.toUpperCase().padEnd(5)} ${result.sample} ${result.profile}`);
  for(const error of result.errors) console.log(`      - ${error}`);
}
console.log(`SUMMARY promoted=${required.length} checks=${results.length} failures=${failures.length}`);
process.exit(failures.length?1:0);

File: notale-v2/experiments/workflow-skills-next/sample-workbench/review/interaction-general/randomness-lab/tests/browser_check.py
#!/usr/bin/env python3
import asyncio,json
from pathlib import Path
from playwright.async_api import async_playwright

ROOT=Path(__file__).resolve().parents[1]
PAGE=(ROOT/'output/pages/index.html').as_uri()
SHOTS=ROOT/'output/.codex-shots'

async def main():
    SHOTS.mkdir(parents=True,exist_ok=True)
    report={'http_requests':[],'errors':[]}
    async with async_playwright() as p:
        browser=await p.chromium.launch(args=['--allow-file-access-from-files'])
        context=await browser.new_context(viewport={'width':1600,'height':900})
        page=await context.new_page()
        page.on('pageerror',lambda e:report['errors'].append(str(e)))
        page.on('request',lambda r:report['http_requests'].append(r.url) if r.url.startswith(('http:','https:')) else None)
        await page.route('http://**/*',lambda route:route.abort())
        await page.route('https://**/*',lambda route:route.abort())
        await page.goto(PAGE,wait_until='load')
        for bit in '010101010101':await page.click(f'[data-bit="{bit}"]')
        report['attempt1']=await page.evaluate('window.__RANDOMNESS_LAB__.getState()')
        await page.wait_for_timeout(800)
        await page.screenshot(path=str(SHOTS/'attempt-1-1600.png'))
        await page.click('#retry')
        for bit in '011010011101':await page.keyboard.press('h' if bit=='1' else 't')
        report['attempt2']=await page.evaluate('window.__RANDOMNESS_LAB__.getState()')
        report['result_text']=await page.locator('.evidence').inner_text()
        await page.wait_for_timeout(800)
        await page.screenshot(path=str(SHOTS/'attempt-2-1600.png'))
        await page.click('#reset');initial1=await page.evaluate('window.__RANDOMNESS_LAB__.getState()')
        await page.click('[data-bit="1"]');await page.click('[data-bit="0"]');await page.click('#reset')
        initial2=await page.evaluate('window.__RANDOMNESS_LAB__.getState()')
        report['reset_equal']=initial1==initial2
        scales=[]
        for w,h in ((1280,720),(1920,1080),(1600,900)):
            await page.set_viewport_size({'width':w,'height':h});await page.evaluate('Deck.resize()');await page.wait_for_timeout(120)
            scales.append(await page.evaluate("""(()=>{const r=document.querySelector('#stage').getBoundingClientRect();return {viewport:[innerWidth,innerHeight],scale:Deck.s,stage:[r.x,r.y,r.width,r.height],scroll:[document.documentElement.scrollWidth,document.documentElement.scrollHeight],overflow:getComputedStyle(document.body).overflow}})()"""))
        report['resize']=scales
        await context.close()
        reduced=await browser.new_context(viewport={'width':1600,'height':900},reduced_motion='reduce')
        rp=await reduced.new_page();await rp.goto(PAGE,wait_until='load')
        for bit in '011010011101':await rp.keyboard.press('h' if bit=='1' else 't')
        report['reduced_motion']=await rp.evaluate("({requested:Deck.reduced(),phase:__RANDOMNESS_LAB__.getState().phase,scanner:getComputedStyle(document.querySelector('#scanner')).animationDuration})")
        await reduced.close();await browser.close()
    assert report['attempt1']['phase']==report['attempt2']['phase']=='result'
    assert report['attempt2']['previous']['sequence']==report['attempt1']['sequence']
    assert report['reset_equal'] and not report['http_requests'] and not report['errors']
    assert all(max(abs(a-b) for a,b in zip(x['stage'],[0,0,*x['viewport']]))<.01 and x['overflow']=='hidden' for x in report['resize'])
    assert report['reduced_motion']['requested'] and report['reduced_motion']['phase']=='result'
    (ROOT/'output/browser-check.json').write_text(json.dumps(report,ensure_ascii=False,indent=2),encoding='utf-8')
    print(json.dumps(report,ensure_ascii=False,indent=2))

asyncio.run(main())
